servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpAutoDraining.java

```diff
@@ -0,0 +1,130 @@
+package io.servicetalk.http.netty;
+
+import io.servicetalk.concurrent.Cancellable;
+import io.servicetalk.concurrent.SingleSource;
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.concurrent.api.TerminalSignalConsumer;
+import io.servicetalk.concurrent.internal.CancelImmediatelySubscriber;
+import io.servicetalk.http.api.HttpExecutionStrategy;
+import io.servicetalk.http.api.HttpServiceContext;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+import io.servicetalk.http.api.StreamingHttpResponseFactory;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.http.api.StreamingHttpServiceFilterFactory;
+
+import javax.annotation.Nullable;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static io.servicetalk.concurrent.api.Single.defer;
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static io.servicetalk.http.api.HttpExecutionStrategies.offloadNone;
+
+public final class HttpAutoDraining implements StreamingHttpServiceFilterFactory {
+
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return offloadNone();
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(StreamingHttpService service) {
+        return new DrainingStreamingHttpServiceFilter(service);
+    }
+
+    private static final class DrainingStreamingHttpServiceFilter extends StreamingHttpServiceFilter {
+
+        DrainingStreamingHttpServiceFilter(StreamingHttpService service) {
+            super(service);
+        }
+
+        @Override
+        public Single<StreamingHttpResponse> handle(HttpServiceContext ctx, StreamingHttpRequest request,
+                                                    StreamingHttpResponseFactory responseFactory) {
+            return defer(() -> doHandle(ctx, request, responseFactory).shareContextOnSubscribe());
+        }
+
+        private Single<StreamingHttpResponse> doHandle(HttpServiceContext ctx, StreamingHttpRequest request,
+                                                    StreamingHttpResponseFactory responseFactory) {
+            final AtomicBoolean payloadSubscribed = new AtomicBoolean();
+            request.payloadBody().beforeOnSubscribe(ignored -> payloadSubscribed.set(true));
```

Comment:
1. Think doesn't work because the returned publisher with applied operator is dropped and not chained. We should use `transformMessageBody`.
2. `beforeOnSubscribe` is slightly delayed signal. Consider using `beforeSubscriber` to set the flag inside supplier that will return a NOOP Subscriber. That way you can intercept `handleSubscribe` call that goes up instead of `onSubscribe` call that goes down in response to subscribe operation.